src/main/java/org/dcsa/ebl/service/impl/TransportDocumentTOServiceImpl.java

```diff
@@ -1,28 +1,207 @@
+
+        return Flux.concat(
+                transportDocumentService.findById(transportDocumentID)
+                    .flatMap(
+                            transportDocument -> {
+                                MappingUtil.copyFields(
+                                        transportDocument,
+                                        transportDocumentTO,
+                                        AbstractTransportDocument.class
+                                );
+                                if (transportDocument.getShippingInstructionID() == null) {
+                                    return Mono.error(new GetException("No ShippingInstruction connected to this TransportDocument - internal error!"));
+                                } else {
+                                    return Flux.concat(
+                                            shippingInstructionTOService.findById(transportDocument.getShippingInstructionID())
+                                                    .flatMap(
+                                                            shippingInstruction -> {
+                                                                transportDocumentTO.setShippingInstruction(shippingInstruction);
+                                                                String carrierBookingReference = null;
+                                                                if (shippingInstruction.getCarrierBookingReference() != null) {
+                                                                    // Use the carrierBookingReference on the ShippingInstruction
+                                                                    carrierBookingReference = shippingInstruction.getCarrierBookingReference();
+                                                                } else {
+                                                                    List<CargoItemTO> cargoItems = shippingInstruction.getCargoItems();
+                                                                    if (cargoItems != null) {
+                                                                        for (CargoItemTO cargoItem : cargoItems) {
+                                                                            if (cargoItem.getCarrierBookingReference() != null) {
+                                                                                // Assume https://github.com/dcsaorg/DCSA-EBL/issues/56 is valid
+                                                                                carrierBookingReference = cargoItem.getCarrierBookingReference();
+                                                                                break;
+                                                                            }
+                                                                        }
+                                                                    }
+                                                                }
```

Comment:
I think it might make sense to move this logic to ShippingInstructionTO (or at least out of the method there - it is getting hard to keep track of what is happening)